File: java/src/main/java/paper/pss/exp/jackson_project/metamorphicRelations/MR3_relation.java
package paper.pss.exp.jackson_project.metamorphicRelations;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import paper.pss.exp.jackson_project.model.MetamorphicRelation;
import paper.pss.exp.jackson_project.model.MetamorphicGroup;
import paper.pss.exp.jackson_project.model.TestCase;

/**
 * MR3: Truncation Relation
 * Tests whether removing digits from the end of a number produces the expected result
 * based on integer division by powers of 10
 */
public class MR3_relation implements MetamorphicRelation {
    private static final Random random = new Random();

    @Override
    public String getId() {
        return "MR3";
    }

    @Override
    public String getDescription() {
        return "Truncation Relation - Tests removing digits from the end";
    }

    @Override
    public List<TestCase> generateFollowupTests(TestCase sourceTest) {
        List<TestCase> followupTests = new ArrayList<>();
        String inputString = sourceTest.getInput();

        if (inputString == null || inputString.isEmpty()) {
            return followupTests;
        }

        try {
            // Check if it's a valid number
            Integer.parseInt(inputString);

            // Skip leading minus sign for truncation
            int startIndex = 0;
            boolean isNegative = false;
            if (inputString.startsWith("-")) {
                startIndex = 1;
                isNegative = true;
            }

            // If only 1 digit, cannot truncate
            if (inputString.length() - startIndex <= 1) {
                return followupTests;
            }

            // Generate 1-3 different truncation lengths
            int maxTruncations = Math.min(3, inputString.length() - startIndex - 1);
            List<Integer> truncationLengths = new ArrayList<>();

            for (int i = 0; i < maxTruncations; i++) {
                // Generate a truncation length (remove 1 to n-1 digits)
                int length;
                do {
                    length = 1 + random.nextInt(inputString.length() - startIndex - 1);
                } while (truncationLengths.contains(length));

                truncationLengths.add(length);

                // Create follow-up test by removing 'length' digits from the end
                String truncatedString = inputString.substring(0, inputString.length() - length);

                // Ensure the follow-up is a valid number
                try {
                    Integer.parseInt(truncatedString);
                    followupTests.add(new TestCase(truncatedString, sourceTest.getPartitionId()));
                } catch (NumberFormatException e) {
                    // Skip invalid follow-up tests
                }
            }
        } catch (NumberFormatException e) {
            // Not applicable if the source is not a valid number
        }

        return followupTests;
    }

    @Override
    public boolean verifyRelation(TestCase sourceTest, TestCase followupTest, int sourceResult, int followupResult, String sourceExecution, String followupExecution) {
        // Get input strings
        String sourceString = sourceTest.getInput();
        String followupString = followupTest.getInput();

        // Calculate how many digits were removed
        int digitsRemoved = sourceString.length() - followupString.length();

        // If no digits were removed, the results should be equal
        if (digitsRemoved == 0) {
            return sourceResult == followupResult;
        }

        // Calculate the expected follow-up result by truncating the source result
        long expectedFollowup = sourceResult / (long) Math.pow(10, digitsRemoved);

        // Check for integer overflow
        if (expectedFollowup > Integer.MAX_VALUE || expectedFollowup < Integer.MIN_VALUE) {
            return false;
        }

        // Compare the actual follow-up result with the expected result
        return followupResult == (int) expectedFollowup;
    }

    @Override
    public boolean isApplicableTo(TestCase testCase) {
        String inputString = testCase.getInput();

        // Not applicable to null inputs
        if (inputString == null) {
            return false;
        }

        try {
            // Check if it's a valid number
            Integer.parseInt(inputString);

            // Skip leading minus sign for length calculation
            int startIndex = inputString.startsWith("-") ? 1 : 0;

            // Only applicable if there are at least 2 digits to truncate
            return inputString.length() - startIndex >= 2;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    @Override
    public List<MetamorphicGroup> createGroups(TestCase sourceTest) {
        List<MetamorphicGroup> groups = new ArrayList<>();
        List<TestCase> followupTests = generateFollowupTests(sourceTest);

        for (TestCase followupTest : followupTests) {
            groups.add(new MetamorphicGroup(
                    getId(),
                    getDescription(),
                    sourceTest,
                    followupTest));
        }

        return groups;
    }

    /**
     * 测试和调试MR3关系的主方法
     */
    public static void main(String[] args) {
        System.out.println("=== MR3 截断关系测试 ===");
        MR3_relation mr3 = new MR3_relation();
        
        // 测试用例
        String[] testInputs = {"12345", "987654321", "123", "12", "1", "0", "-12345", "abc123", "", null};
        
        for (String input : testInputs) {
            System.out.println("\n测试输入: " + (input == null ? "null" : "\"" + input + "\""));
            
            TestCase sourceTest = new TestCase(input, 1);
            
            // 检查是否适用
            boolean applicable = mr3.isApplicableTo(sourceTest);
            System.out.println("  适用性: " + applicable);
            
            if (!applicable) {
                continue;
            }
            
            // 生成后续测试
            List<TestCase> followupTests = mr3.generateFollowupTests(sourceTest);
            System.out.println("  生成的后续测试数量: " + followupTests.size());
            
            for (TestCase followup : followupTests) {
                System.out.println("    后续测试: \"" + followup.getInput() + "\"");
                
                // 尝试验证关系
                try {
                    int sourceResult = Integer.parseInt(input);
                    int followupResult = Integer.parseInt(followup.getInput());
                    
                    boolean verified = mr3.verifyRelation(sourceTest, followup, sourceResult, followupResult, "", "");
                    
                    System.out.println("      源结果: " + sourceResult + ", 后续结果: " + followupResult);
                    System.out.println("      关系验证: " + (verified ? "✓ 通过" : "✗ 失败"));
                    
                    // 验证截断逻辑
                    String expectedTruncation = input.substring(0, input.length() - 1);
                    System.out.println("      预期截断: \"" + expectedTruncation + "\"");
                    System.out.println("      实际后续: \"" + followup.getInput() + "\"");
                    System.out.println("      截断匹配: " + (expectedTruncation.equals(followup.getInput()) ? "✓" : "✗"));
                    
                    // 验证数学关系
                    int expectedResult = sourceResult / 10;
                    System.out.println("      数学验证: " + sourceResult + " / 10 = " + expectedResult);
                    System.out.println("      结果匹配: " + (expectedResult == followupResult ? "✓" : "✗"));
                    
                } catch (NumberFormatException e) {
                    System.out.println("      解析错误，测试错误处理");
                    boolean verified = mr3.verifyRelation(sourceTest, followup, 0, 0, "NumberFormatException", "NumberFormatException");
                    System.out.println("      错误处理验证: " + (verified ? "✓ 通过" : "✗ 失败"));
                }
            }
            
            // 测试蜕变组创建
            List<MetamorphicGroup> groups = mr3.createGroups(sourceTest);
            System.out.println("  创建的蜕变组数量: " + groups.size());
        }
        
        System.out.println("\n=== MR3测试完成 ===");
    }
}

File: python/incomeTax/metamorphicRelations/mr3_relation.py
import random
from typing import List

from python.incomeTax.model.test_case import TestCase
from python.incomeTax.model.metamorphic_relation import MetamorphicRelation

# 定义最高收入和最低收入
EXEMPTION_AMOUNT = 5000.0
MAX_INCOME = 1400000


class MR3Relation(MetamorphicRelation):
    """
    MR3: 输入缩放关系

    验证当收入按比例增加时，所得税也应该相应增加。
    """

    def get_id(self) -> str:
        """获取蜕变关系ID"""
        return "mr_3"

    def get_description(self) -> str:
        """获取蜕变关系描述"""
        return "输入缩放关系 - 验证当收入按比例增加时，所得税也应该相应增加"

    def generate_followup_tests(self, source_test: TestCase) -> List[TestCase]:
        """
        基于源测试用例生成后续测试用例

        创建一个收入按比例缩放的后续测试用例。

        Args:
            source_test: 源测试用例

        Returns:
            包含一个后续测试用例的列表
        """
        followup_tests = []

        income_value = source_test.get_income_value()
        # 缩放因子，对应IncomeTax.py中的mr_3
        scaling_factor1 = random.uniform(1, 2)
        scaling_factor2 = random.uniform(1, 2)

        new_income_value1 = min(income_value * scaling_factor1, MAX_INCOME)
        new_income_value2 = min(income_value * scaling_factor2, MAX_INCOME)

        followup_test1 = TestCase(new_income_value1)
        followup_test2 = TestCase(new_income_value2)
        followup_tests.extend([followup_test1, followup_test2])

        return followup_tests

    def verify_relation(
        self,
        source_test: TestCase,
        followup_test: TestCase,
        source_result: float,
        followup_result: float,
        source_execution: str = "",
        followup_execution: str = "",
    ) -> bool:
        """
        验证源测试和后续测试的结果是否满足蜕变关系

        验证后续测试的税额是否大于或等于源测试的税额。

        Args:
            source_test: 源测试用例
            followup_test: 后续测试用例
            source_result: 源测试结果
            followup_result: 后续测试结果
            source_execution: 源测试执行的错误信息，如果有的话
            followup_execution: 后续测试执行的错误信息，如果有的话

        Returns:
            如果满足输入缩放关系则返回True
        """
        # 检查是否有任何执行错误
        if source_execution or followup_execution:
            # 如果两个测试都有相同类型的错误，关系仍被认为是满足的
            return source_execution == followup_execution

        # 简单的缩放关系，税额应该随收入增加而增加
        return followup_result >= source_result - 1e-6

    def is_applicable_to(self, test_case: TestCase) -> bool:
        """
        检查蜕变关系是否适用于给定的测试用例

        这个关系适用于收入大于等于免征额，且有缩放空间的测试用例。

        Args:
            test_case: 要检查的测试用例

        Returns:
            如果测试用例的收入大于等于免征额且缩放后不超过最高收入则返回True，否则返回False
        """
        # 确保有足够的空间进行收入缩放
        # 假设最小缩放因子为1.01，确保缩放后不超过MAX_INCOME
        return True


if __name__ == "__main__":
    mr = MR3Relation()
    print(f"MR ID: {mr.get_id()}")
    print(f"MR Description: {mr.get_description()}")

    # 示例测试用例
    source_test = TestCase(income_value=10000.0)
    print(f"\nSource Test Case Income: {source_test.get_income_value()}")

    # 生成后续测试用例
    followup_tests = mr.generate_followup_tests(source_test)
    print("Follow-up Test Cases Income:")
    for ft in followup_tests:
        print(f"  {ft.get_income_value()}")

    # 模拟结果并验证关系
    source_result = 500.0  # 假设的税额
    print(f"\nSource Result: {source_result}")

    for ft in followup_tests:
        # 模拟一个增加的税额
        followup_result = source_result * random.uniform(
            1.0, 1.5
        )  # 假设税额随收入缩放而增加
        is_satisfied = mr.verify_relation(
            source_test, ft, source_result, followup_result
        )
        print(
            f"  Follow-up Income: {ft.get_income_value()}, Result: {followup_result}, Satisfied: {is_satisfied}"
        )

    # 测试不适用情况
    source_test_not_applicable_low = TestCase(income_value=4000.0)  # 低于免征额
    print(
        f"\nIs applicable to {source_test_not_applicable_low.get_income_value()} (below exemption): {mr.is_applicable_to(source_test_not_applicable_low)}"
    )

    source_test_not_applicable_high = TestCase(income_value=990000.0)  # 接近最高收入
    print(
        f"Is applicable to {source_test_not_applicable_high.get_income_value()} (near max income): {mr.is_applicable_to(source_test_not_applicable_high)}"
    )
